### PR TITLE
 Add volumesnapshot content deletion policy

### PR DESCRIPTION
**Pull request by jingxu97**
This PR adds DeletionPolicy which describes a policy for end-of-life maintenance of volume snapshot contents.
There are two types of policies, delete and retain. By default, the
policy is delete. In this case, when volume snaphsot is deleted, the
volume snapshot content for it should be deleted, as well as the
physical snapshot. If the policy is set the retain, the volume snapshot
content will remain even if the volume snapshot it binds to is deleted.

### CODE REVIEW

File: pkg/controller/snapshot_controller.go [hunk 1]
```diff
@@ -125,9 +125,21 @@ func (ctrl *csiSnapshotController) syncContent(content *crdv1.VolumeSnapshotCont
...
+		if content.Spec.DeletionPolicy != nil {
+			switch *content.Spec.DeletionPolicy {
+			case crdv1.VolumeSnapshotContentRetain:
+				glog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Retain, nothing to do", content.Name)
+		
+			case crdv1.VolumeSnapshotContentDelete:
+				glog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Delete", content.Name)
+				ctrl.deleteSnapshotContent(content)
+			default:
+				// Unknown PersistentVolumeReclaimPolicy
```
Comment (xing-yang) [inline pkg/controller/snapshot_controller.go]: s/PersistentVolumeReclaimPolicy/VolumeSnapshotReclaimPolicy

File: pkg/controller/snapshot_controller.go [hunk 2]
```diff
@@ -125,9 +125,21 @@ func (ctrl *csiSnapshotController) syncContent(content *crdv1.VolumeSnapshotCont
...
+				ctrl.eventRecorder.Event(content, v1.EventTypeWarning, "SnapshotUnknownDeletionPolicy", "Volume Snapshot Content has unrecognized deletion policy")
```
Comment (xing-yang) [inline pkg/controller/snapshot_controller.go]: So the default policy is actually the same as Retain?

File: pkg/controller/snapshot_delete_test.go [hunk 3]
```diff
@@ -128,10 +129,21 @@ func TestDeleteSync(t *testing.T) {
 			expectedDeleteCalls: []deleteCall{{"sid1-1", nil, nil}},
 			test:                testSyncContent,
 		},
+		{
+			name:                "1-1 - content with retain policy will not be deleted if it is bound to a non-exist snapshot and also has a snapshot uid specified",
```
Comment (wackxu) [inline pkg/controller/snapshot_delete_test.go]: the number of test case is the same as bellow, we can put this and "2-1" case after the "1-11" case and rename it "1-12"

File: pkg/controller/snapshot_controller.go (hunk 1 above)
Comment (jingxu97) [reply pkg/controller/snapshot_controller.go]: fixed

File: pkg/controller/snapshot_controller.go (hunk 2 above)
Comment (jingxu97) [reply pkg/controller/snapshot_controller.go]: no, it should be Delete. I fixed the logic here.

File: pkg/apis/volumesnapshot/v1alpha1/types.go [hunk 4]
```diff
@@ -241,3 +249,15 @@ type CSIVolumeSnapshotSource struct {
...
+
+// DeletionPolicy describes a policy for end-of-life maintenance of volume snapshot contents
+type DeletionPolicy string
+
+const (
+	// VolumeSnapshotContentDelete means the snapshot content will be deleted from Kubernetes on release from its volume snapshot.
+	// The volume plugin must support Deletion.
+	VolumeSnapshotContentDelete DeletionPolicy = "Delete"
+	// VolumeSnapshotContentRetain means the volume will be left in its current phase (Released) for manual reclamation by the administrator.
+	// The default policy is Retain.
```
Comment (wackxu) [inline pkg/apis/volumesnapshot/v1alpha1/types.go]: update the comment here since the default policy `Delete`

Comment (jingxu97) [reply pkg/apis/volumesnapshot/v1alpha1/types.go]: Sorry actually here we should set default policy for VolumeSnapshotContent is Retain.
Similar to Persistent Volumes, if VolumeSnapshotContent is created manually, the default set is Retain. If VolumeSnapshotContent is dynamically created through VolumeSnapshot, the default policy is Delete which is set through VolumeSnaphshotClass.

**Comment by wackxu** [comment]
The CI is failed, could you update failed UT?

File: pkg/controller/snapshot_delete_test.go (hunk 3 above)
Comment (jingxu97) [reply pkg/controller/snapshot_delete_test.go]: fixed

**Comment by xing-yang** [comment]
@jingxu97 There's CI failure.  Can you check?

File: pkg/controller/snapshot_controller.go (hunk 2 above)
Comment (xing-yang) [reply pkg/controller/snapshot_controller.go]: For default, it won't call deleteSnapshotContent.  So default is still equal to Retain?  You said it should be the same as Delete earlier.

Comment (jingxu97) [reply pkg/controller/snapshot_controller.go]: there are two cases. 
For dynamic creation, we add deletePolicy when creating snapshotContent. Check "createSnapshotOperation" function in controller

For static binding case, if no deletePolicy is set, by default it is retain.

**Comment by wackxu** [comment]
LGTM

**Comment by xing-yang** [comment]
Looks good now.  Will merge this as soon as RC3 is cut.

**Pull request merged**